proxy.go

```diff
@@ -236,28 +236,28 @@ func executeWithRetry(
 		// StatusBadGateway response is returned by http.ReverseProxy when
 		// it cannot establish connection to remote host.
 		if rw.StatusCode() == http.StatusBadGateway {
-			log.Debugf("the invalid host is: %s", s.host.addr)
-			s.host.penalize()
+			log.Debugf("the invalid host is: %s", s.host)
+			s.host.Penalize()
 			// comment s.host.dec() line to avoid double increment; issue #322
 			// s.host.dec()
-			atomic.StoreUint32(&s.host.active, uint32(0))
-			newHost := s.host.replica.cluster.getHost()
+			s.host.SetIsActive(false)
+			newHost := s.cluster.getHost()
```

Comment:
scope has a reference to cluster, so no need to resolve the cluster via host->replica->cluster.